FreePIE.Core.Plugins/joystick/JoystickGlobal.cs

```diff
@@ -13,152 +13,193 @@ public class JoyProp
 
-        public float x
-        {
-            get 
+        public JoyCount count => new JoyCount
             {
-                return normalize(State.X, JoystickOffset.X);
-            }
-        }
+                povs = _joystick.Capabilities.PovCount,
+                buttons = _joystick.Capabilities.ButtonCount,
+                axes = _joystick.Capabilities.AxeCount,
+                sliders = _state.Sliders.Length//.Where(s => s != 0 ).Count()
+            };
 
-        public float y
-        {
-            get { return normalize(State.Y, JoystickOffset.Y); }
-        }
 
-        public float z
-        {
-            get { return normalize(State.Z, JoystickOffset.Z); }
-        }
+        public JoyConfig config { get; private set; }
 
-        public float rotationX
-        {
-            get { return normalize(State.RotationX, JoystickOffset.RotationX); }
-        }
+        public event EventHandler<JoystickEventArgs> Updated;
 
-        public float rotationY
-        {
-            get { return normalize(State.RotationY, JoystickOffset.RotationZ); }
-        }
+        public event EventHandler Started, Stopped;
 
-        public float rotationZ
-        {
-            get { return normalize(State.RotationZ, JoystickOffset.RotationZ); }
-        }
 
+        public double x => normalize(_state?.X??0, config.x );
+        public double y => normalize(_state?.Y ?? 0, config.y);
+        public double z => normalize(_state?.Z ?? 0, config.z);
+        public double rotationX => normalize(_state?.RotationX ?? 0, config.rotationX);
+        public double rotationY => normalize(_state?.RotationY ?? 0, config.rotationY);
+        public double rotationZ => normalize(_state?.RotationZ ?? 0, config.rotationZ); 
+            
         
 
         public bool[] buttons { get; private set; }
 
-        public float[] sliders
+        public double[] sliders
         {
-            get { return State.Sliders.Select((s,i) => normalize(s, i == 0 ? JoystickOffset.Sliders0 : JoystickOffset.Sliders1 )).ToArray(); }
+            get { return _state.Sliders.Take(count.sliders).Select((s,i) => normalize(s, config.sliders[i])).ToArray(); }
         }
 
         public int[] pov
         {
             get
             {
-                return numPovs > 0 ? State.PointOfViewControllers.Select(p =>
+                return _joystick.Capabilities.PovCount > 0 ? _state.PointOfViewControllers.Take(count.povs).Select(p =>
                 {
                     var r = p > 0 ? p / 100 : p;
                     return r;
                 }).ToArray() : new[] { -1, -1, -1, -1 };
             }
         }
 
-        private float normalize(int value, JoystickOffset axis )
+
+        public DpadGlobal[] dpad => pov.Select(p => new DpadGlobal(p)).ToArray();
+
+
+
+
+        #endregion
+
+        #region Private Methods
+        private double mapRange(double x, double xMin, double xMax, double yMin, double yMax)
```

Comment:
actually this is a weird one.  I wanted to just use the function from Freepie.Core FilterHelper 
but plugins do not reference Freepie.Core.  So i had to copy it over here for private use,  not meant to be reused.

I could add a reference to Freepie.Core from plugins,  but that would go against what they were going for with the plugin system. (It only references Contracts, so that plugin authors can make their own plugins and just drop them in the freepie folder)

So Long story short these methods are private and will not be reused.

Reply:
Unless you mean to allow this one function to be reused somewhere else within Freepie.Plugin ,  in which case we would be better off creating an interface for FilterHelper and adding it into FreePie contracts. 

That might not work either since Filterhelper isnt actually registered with the Container directly.  All of the Helpers are simply registered as IScriptHelper and resolve as an IEnumerable.  

So we would need to individually register IFilterHelper with the container.   

It gets really complicated.